src/backend/gl/src/device.rs

```diff
@@ -468,16 +469,74 @@ pub(crate) unsafe fn set_sampler_info<SetParamFloat, SetParamFloatVec, SetParamI
 impl d::Device<B> for Device {
     unsafe fn allocate_memory(
         &self,
-        _mem_type: c::MemoryTypeId,
+        mem_type: c::MemoryTypeId,
         size: u64,
     ) -> Result<n::Memory, d::AllocationError> {
-        // TODO
-        Ok(n::Memory {
-            properties: memory::Properties::CPU_VISIBLE | memory::Properties::CPU_CACHED,
-            first_bound_buffer: Cell::new(None),
-            size,
-            emulate_map_allocation: RefCell::new(std::ptr::null_mut()),
-        })
+        if (1 << mem_type.0) & IMAGE_MEM_TYPE_MASK != 0 {
+            assert_ne!((1 << mem_type.0) & DEVICE_LOCAL_MASK, 0);
+            Ok(n::Memory {
+                properties: memory::Properties::DEVICE_LOCAL,
+                buffer: None,
+                target: 0,
+                size,
+                emulate_map_allocation: Cell::new(None),
+                map_flags: 0,
+            })
+        } else {
+            assert!(self.share.private_caps.buffer_role_change);
+
+            let target = if (1 << mem_type.0) & INDEX_MEM_TYPE_MASK != 0 {
+                glow::ELEMENT_ARRAY_BUFFER
+            } else {
+                glow::ARRAY_BUFFER
+            };
+
+            let is_device_local = (1 << mem_type.0) & DEVICE_LOCAL_MASK != 0;
+
+            let gl = &self.share.context;
+
+            let raw = gl.create_buffer().unwrap();
+            //TODO: use *Named calls to avoid binding
+            gl.bind_buffer(target, Some(raw));
+            if self.share.private_caps.buffer_storage {
+                let flags = if is_device_local {
+                    0
+                } else {
+                    glow::MAP_READ_BIT
+                        | glow::MAP_WRITE_BIT
+                        | glow::MAP_PERSISTENT_BIT
+                        | glow::DYNAMIC_STORAGE_BIT
+                };
+                gl.buffer_storage(target, size as i32, None, flags);
+            } else {
+                let usage = if is_device_local {
+                    glow::STATIC_DRAW
+                } else {
+                    glow::DYNAMIC_DRAW
+                };
+                gl.buffer_data_size(target, size as i32, usage);
+            }
+            gl.bind_buffer(target, None);
+
+            if let Err(err) = self.share.check() {
+                panic!("Error allocating memory buffer {:?}", err);
+            }
+
+            let mut map_flags = glow::MAP_READ_BIT | glow::MAP_WRITE_BIT | glow::MAP_PERSISTENT_BIT;
+            if (1 << mem_type.0) & COHERENT_MASK == 0 {
+                map_flags |= glow::MAP_FLUSH_EXPLICIT_BIT;
+            }
+
+            Ok(n::Memory {
+                properties: memory::Properties::CPU_VISIBLE
+                    | memory::Properties::CPU_CACHED,
```

Comment:
in the future, we'd want to expose "read" memory type independent of the "write", but it doesn't need to happen here

Reply:
Yeah, agreed!